llpc/util/llpcElfWriter.cpp

```diff
@@ -735,6 +737,285 @@ template <class Elf> void ElfWriter<Elf>::updateMetaNote(Context *pContext, cons
+  pNewSection->secHead.sh_size = newSectionSize;
+
+  auto data = mergedData;
+
+  memcpy(data, pSection1->data, section1Size);
+  data += section1Size;
+
+  if (pSection2) {
+
+    auto section2Relocs = reinterpret_cast<typename Elf::Reloc *>(const_cast<uint8_t *>(pSection2->data));
+    size_t section2RelocsCount = getRelocsCount(pSection2);
+
+    for (auto i = section2RelocsOffset; i < section2RelocsCount; ++i) {
+
+      addRelocSymbols(reader, section2Relocs[i]);
+
+      // Merge the relocs data
+      typename Elf::Reloc reloc = {};
+      reloc.r_offset = section2Relocs[i].r_offset + diffOfPsOffset;
+      reloc.r_type = section2Relocs[i].r_type;
+      ElfSymbol section2RelocSymbol = {};
+      reader.getSymbol(section2Relocs[i].r_symbol, &section2RelocSymbol);
+      reloc.r_symbol = getRelocSymbolIndex(section2RelocSymbol);
+      memcpy(data, &reloc, sizeof(typename Elf::Reloc));
+      data += sizeof(typename Elf::Reloc);
+    }
+  }
+}
+
+// =====================================================================================================================
+// Process the reloc section
+//
+// @param reader : The ElfReader to input fragment elf bin
+// @param nonFragmentPsIsaOffset : The offset of PS ISA in nonFragment elf bin
+// @param fragmentPsIsaOffset : The offset of PS ISA in input fragment elf bin
+template <class Elf>
+void ElfWriter<Elf>::processRelocSection(const ElfReader<Elf> &reader, size_t nonFragmentPsIsaOffset,
+                                         size_t fragmentPsIsaOffset) {
+
+  ElfSectionBuffer<Elf64::SectionHeader> *fragmentRelocSection = nullptr;
+  const ElfSectionBuffer<Elf64::SectionHeader> *nonFragmentRelocSection = nullptr;
+
+  auto fragmentRelocSecIndex = reader.GetSectionIndex(RelocName);
+  auto nonFragmentRelocSecIndex = GetSectionIndex(RelocName);
+
+  reader.getSectionDataBySectionIndex(fragmentRelocSecIndex, &fragmentRelocSection);
+  getSectionDataBySectionIndex(nonFragmentRelocSecIndex, &nonFragmentRelocSection);
+
+  bool isFragmentRelocUsedInPs = false;
+
+  // The start pos of the relocs which used by PS
+  size_t fragmentRelocPsStartPos = 0;
+
+  if (fragmentRelocSection) {
+    size_t fragmentRelocCounts = getRelocsCount(fragmentRelocSection);
+    fragmentRelocPsStartPos = getRelocPsStartPos(fragmentRelocSection, fragmentPsIsaOffset);
+    isFragmentRelocUsedInPs = (fragmentRelocPsStartPos < fragmentRelocCounts);
+  }
+
+  if (!nonFragmentRelocSection && isFragmentRelocUsedInPs) {
+    // The input fragment elf bin has relocs used in PS, but nonFragment elf bin does not have reloc section,
+    // then a reloc section need to be created for nonFragment elf bin
+    ElfSectionBuffer<Elf64::SectionHeader> newRelocSection = {};
+
+    createNewSection(RelocName, fragmentRelocSection, &newRelocSection);
+
+    m_sections.push_back(newRelocSection);
+    // Update the reloc section index
+    m_map[RelocName] = m_sections.size() - 1;
+    m_relocSecIdx = m_map[RelocName];
```

Comment:
Swapping these two is one less lookup in the map
```cpp
    m_relocSecIdx = m_sections.size() - 1;
    m_map[RelocName] = m_relocSecIdx;
```

Reply:
Addressed.